datafusion/expr/src/expr.rs

```diff
@@ -251,6 +251,77 @@ pub enum Expr {
     QualifiedWildcard { qualifier: String },
 }
 
+/// Recursively find all columns referenced by an expression
+pub fn find_columns_referenced_by_expr(e: &Expr) -> Vec<Column> {
```

Comment:
I think this code could use an ExprVisitor to avoid having to do the recursion itself (and potentially missing some case)

I took a crack at doing so https://github.com/apache/arrow-datafusion/pull/2471